test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,529 @@
+//
+// Copyright (c) 2021 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "testBase.h"
+#include "harness/testHarness.h"
+#include <vector>
+
+/* Negative Tests for clGetDeviceInfo */
+int test_negative_get_device_info(cl_device_id deviceID, cl_context context,
+                                  cl_command_queue queue, int num_elements)
+{
+
+    cl_device_type device_type = 0;
+    cl_int err = clGetDeviceInfo(nullptr, CL_DEVICE_TYPE, sizeof(device_type),
+                                 &device_type, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_DEVICE,
+        "clGetDeviceInfo should return CL_INVALID_DEVICE when: \"device is not "
+        "a valid device\" with a nullptr",
+        TEST_FAIL);
+
+    err =
+        clGetDeviceInfo(reinterpret_cast<cl_device_id>(context), CL_DEVICE_TYPE,
```

Comment:
This spec issue is related, I think: https://github.com/KhronosGroup/OpenCL-Docs/issues/482

Device IDs are slightly different since you "get" them rather than "create" them and they aren't associated with a context, but there are a lot of similarities to the objects discussed on the issue as well.

Reply:
Thanks, that's what I was thinking of. I guess our own discussion has broadened to the more generic question of whether error-checking should be required to distinguish between any two resource objects (specifically those implementing the icd _cl_<object> layout), so it's a question for negative testing on almost every API.